### PR TITLE
[doc] DeviceMesh invariant on DTensorSpec

### PR DESCRIPTION
**Pull request by ezyang**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.11.0) (oldest at bottom):
* __->__ #157806
* #157805

Signed-off-by: Edward Yang <ezyang@meta.com>

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

**Review by wanchaol — APPROVED** [review]

File: torch/distributed/tensor/_op_schema.py [hunk 1]
```diff
@@ -132,6 +134,8 @@ class OpStrategy(StrategyType):
     """
     OpStrategy that consists of a list of sharding strategies associated with the op,
     where each strategy is an OpSpec that describes the acceptable input/output sharding.
+
+    invariant: the DeviceMesh on all OpSpec must be the same
```
Comment (wconstab) [inline torch/distributed/tensor/_op_schema.py:138]: @tianyu-l is working on EP design and there are cases where we encounter different device meshes on different tensors (e.g expert params on EP mesh vs non expert on non EP mesh). I don't think we resolved how to handle this but one option is to. Allow mixing meshes in certain operators like foreach, which would violate this invariant. 

@tianyu-l what's the latest?

Comment (tianyu-l) [reply torch/distributed/tensor/_op_schema.py:138]: In the context of EP, the issue of having mixed meshes on different parameters only show up in fused / foreach optimizer step, and gradient norm clipping. For other parts of EP, they are doing local computations on individual meshes.

For optimizer step, we are working with `TupleStrategy` https://github.com/pytorch/pytorch/pull/157682, so this statement of invariant on `OpStrategy` sounds OK to me.

For gradient norm clipping, I ended up doing one computation for each set of parameters on the same mesh for now, so this statement is still OK. We can revisit my approach, but I feel it may not be worth altering this assumption just for the grad clipping use case.
